### PR TITLE
Feature/mx-1478 mex-model synchronisation

### PR DESCRIPTION
**Pull request by cutoffthetop**
# model changes
- change `Activity.externalAssociate` to `list[OrganizationID | PersonID]` instead of unit
- change `Distribution.license` and `Resource.license` to new vocabulary enum `Licence`
- change `Distribution.modified` to `Timestamp | None` instead of list
- change `Distribution.publisher `list[OrganizationID]` instead of person list
- change `Organization.wikidataId` to be a list instead of optional string
- add a default of an empty list to `Person.fullName`
- change `Resource.created` and `Resource.modified` to be an optional `Timestamp` instead of list
- change `Resource.stateOfDataProcessing` to be a list instead of optional
- change `Resource.temporal` to be an optional string or Timestamp instead of a list
- update `EMAIL_PATTERN` to be aligned with model
- move `ActivityType`, `AnonymizationPseudonymization`, `APIType`, `DataProcessingState`, `DataType`, `Frequency`, `Language`, `License`, `MIMEType`, `ResourceTypeGeneral` and `TechnicalAccessibility` to `mex.common.types`
- always annotate the types with `Annotated[TYPE, Field()]` rather than setting `= Field()` as default

# schema changes
- add `{"format": "uri"}` to json schema of URLs with regex pattern
- add `{"format": "date-time"}` to Timestamp fields

# common changes
- add dev-dependency to https://github.com/robert-koch-institut/mex-model
- add `dromedary_to_kebab` case converter utility

# testing changes
- add tests to verify `mex.common.models` and `mex-model:schema` are in sync

# misc changes
- update dependencies
- skip `public_api` tests

### CODE REVIEW

File: tests/models/test_schema.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+) -> None:
+    assert set(generated["required"]) == set(specified["required"])
+
+
+@pytest.mark.parametrize(
+    ("entity_type", "field_name"),
+    ENTITY_TYPES_AND_FIELD_NAMES_BY_FQN.values(),
+    ids=ENTITY_TYPES_AND_FIELD_NAMES_BY_FQN.keys(),
+)
+def test_field_defs_match_spec(entity_type: str, field_name: str) -> None:  # noqa: C901
```
Comment (erichesse) [inline tests/models/test_schema.py:72]: I tested it and I can see what it's doing, but unfortunately I did not really crack the function as whole

**Review by erichesse — COMMENTED** [review]
I reviewed the changes short of the test function test_field_defs_match_spec().

As I will be on vacation now, I hope someone else will take over the review. Short of a small issue in a doc string the PR looks good though.

File: tests/models/test_schema.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            name.removeprefix("Extracted"): model.model_json_schema()
+            for name, model in EXTRACTED_MODEL_CLASSES_BY_NAME.items()
+        }.items()
+    )
+)
+SPECIFIED_SCHEMAS = dict(
+    sorted(
+        {
+            schema["title"].replace(" ", ""): schema
+            for file_name in glob.glob(str(SPECIFIED_SCHEMA_PATH / "*.json"))
```
Comment (rababerladuseladim) [inline tests/models/test_schema.py:28]:
```suggestion
            for file_name in SPECIFIED_SCHEMA_PATH.glob("*.json")
```

File: tests/models/test_schema.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+GENERATED_SCHEMAS = dict(
+    sorted(
+        {
+            name.removeprefix("Extracted"): model.model_json_schema()
+            for name, model in EXTRACTED_MODEL_CLASSES_BY_NAME.items()
+        }.items()
+    )
+)
+SPECIFIED_SCHEMAS = dict(
+    sorted(
```
Comment (rababerladuseladim) [inline tests/models/test_schema.py:25]: why sort this?

File: tests/models/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                    e
+                    for item in obj["items"]
+                    for e in (
+                        item.pop("examples", []) if isinstance(item, dict) else []
+                    )
+                ]
+        if "anyOf" in obj:  # ignore any/one difference
+            obj["oneOf"] = obj.pop("anyOf")
+        for quantifier in ("oneOf", "allOf"):
+            if quantifier in obj:
```
Comment (rababerladuseladim) [inline tests/models/test_schema.py:135]: Not sure if this is easy enough to read but at least on less level of indentation 😝 

```suggestion
        for quantifier in ({"oneOf", "allOf"} & set(obj)):
```

File: tests/test_transform.py [hunk 5]
```diff
@@ -93,3 +98,25 @@ def test_snake_to_dromedary(string: str, expected: str) -> None:
...
     assert result == expected
+
+
+@pytest.mark.parametrize(
+    ("string", "expected"),
+    [
+        ("", ""),
+        ("word", "word"),
+        ("already-kebab", "already-kebab"),
+        ("ABWordCDEWordFG", "ab-word-cde-word-fg"),
```
Comment (rababerladuseladim) [inline tests/test_transform.py:109]: Maybe add a case for `Such-AWeird-EdgeCase`?

**Review by rababerladuseladim — CHANGES_REQUESTED** [review]
Ouch, that test is ugly but: An awesome helper, thank your very much for taking up the struggle and building it.

FYI: pydantic includes functions for case converting, although they call them differently and lack kebab: https://docs.pydantic.dev/2.0/api/alias_generators/

Regarding `test_schema`: holy shmoly, these are many edge cases which lead to slight discomfort on my side: We might use this test as an excuse for skipping manual validation of spec and code model concordance in the future. Therefore, I'd like to have this test as airtight as possible with as few ignored discrepancies as possible. What do you think about collecting ideas for reducing discrepancies in a new ticket?

Extractor tests fail with this branch. Are you fixing that in this ticket or will you create a separate?

Did Franzi review data as stated in the ACs or will you move that into a separate extractor-update ticket?

File: tests/test_transform.py (hunk 5 above)
Comment (cutoffthetop) [reply tests/test_transform.py:111]: done [c2d6669](https://github.com/robert-koch-institut/mex-common/pull/48/commits/c2d6669d09cf6aa3b5e0d46b479cd3de7d7d5cae)

File: tests/models/test_schema.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        if field_name == "temporal":  # this field's examples are too complicated
+            obj.pop("examples", None)
+        if obj.get("format") == "date":  # we mash up dates and datetimes in models
+            obj["format"] = "date-time"
+        if obj.get("pattern") in (
+            r"^[1-9]\d{3}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$",
+            r"^\d{4}(-\d{2})?(-\d{2})?$",  # ignore date/time pattern differences
+        ):
+            obj["pattern"] = r"^\d{4}(-\d{2}(-\d{2}(T\d{2}:\d{2}:\d{2}Z)?)?)?$"
+        if obj.get("pattern") == MEX_ID_PATTERN:  # align concept/enum annotations
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply tests/models/test_schema.py]: explained here 30634899d91f5f81c7cff6dea69eca0764e0ae3a

File: tests/models/test_schema.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            obj.pop("type")
+        if obj.get("examples") == [str(Identifier.generate(seed=42))]:  # not in spec
+            obj.pop("examples")
+        if field_name == "temporal":  # this field's examples are too complicated
+            obj.pop("examples", None)
+        if obj.get("format") == "date":  # we mash up dates and datetimes in models
+            obj["format"] = "date-time"
+        if obj.get("pattern") in (
+            r"^[1-9]\d{3}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$",
+            r"^\d{4}(-\d{2})?(-\d{2})?$",  # ignore date/time pattern differences
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply tests/models/test_schema.py]: new comment here https://github.com/robert-koch-institut/mex-common/commit/e8d2f8b4d64667f65e513c4b8d35311de3ee9971#diff-81806985f5f6ddaa66702cdd6cc09346d823d0236091efada21ffe6c0327ae2eR91-R92

File: tests/models/test_schema.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            return
+        obj.pop("sameAs", None)  # not in models
+        obj.pop("subPropertyOf", None)  # not in models
+        obj.pop("description", None)  # not in spec
+        obj.pop("default", None)  # not in spec
+        if obj.get("type") == "null":  # pydantic 2 sets these
+            obj.pop("type")
+        if obj.get("examples") == [str(Identifier.generate(seed=42))]:  # not in spec
+            obj.pop("examples")
+        if field_name == "temporal":  # this field's examples are too complicated
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply tests/models/test_schema.py]: this is gone

File: tests/models/test_schema.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    assert set(generated["required"]) == set(specified["required"])
+
+
+@pytest.mark.parametrize(
+    ("entity_type", "field_name"),
+    ENTITY_TYPES_AND_FIELD_NAMES_BY_FQN.values(),
+    ids=ENTITY_TYPES_AND_FIELD_NAMES_BY_FQN.keys(),
+)
+def test_field_defs_match_spec(entity_type: str, field_name: str) -> None:  # noqa: C901
+    def prepare_field(obj: list[Any] | dict[str, Any]) -> None:  # noqa: C901
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply tests/models/test_schema.py]: cleaned it up a bit in multiple commits, please take a fresh look a the diff

File: tests/models/test_schema.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+            if isinstance(obj["items"], dict):
+                obj["items"] = [obj["items"]]
+            if scheme := obj.pop("useScheme", None):  # pull down scheme into items
+                for item in obj["items"]:
+                    item["useScheme"] = scheme
+            prepare_field(obj["items"])
+            if any(
+                item.get("examples") if isinstance(item, dict) else None
+                for item in obj["items"]
+            ):  # pull up examples onto field level
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply tests/models/test_schema.py]: was a bit of work, but done here -> https://github.com/robert-koch-institut/mex-model/pull/8

**Comment by cutoffthetop** [comment]
> FYI: pydantic includes functions for case converting, although they call them differently and lack kebab: https://docs.pydantic.dev/2.0/api/alias_generators/

yeah, that's a shame about the kebab. maybe we could make a contrib PR in a mob session? otherwise i'd leave it all in `mex.common.transform` for now

File: tests/public_api/test_connector.py [hunk 11]
```diff
@@ -133,6 +133,7 @@ def test_post_models_mocked(
     assert payload == expected_payload
 
 
+@pytest.mark.skip
```
Comment (rababerladuseladim) [inline tests/public_api/test_connector.py:136]:
```suggestion
@pytest.mark.skip(reason="lacking ssl certificate")
```

File: tests/public_api/test_connector.py [hunk 12]
```diff
@@ -318,6 +319,7 @@ def test_get_all_items_mocked(
     assert items == mex_metadata_items_response
 
 
+@pytest.mark.skip
```
Comment (rababerladuseladim) [inline tests/public_api/test_connector.py:322]:
```suggestion
@pytest.mark.skip(reason="lacking ssl certificate")
```

**Review by rababerladuseladim — CHANGES_REQUESTED** [review]
Neat, looks like you also upped your pydantic v2 skillz!

Aaaaand: Working extractors are part of the ticket ACs. Are you on that?

`tests/public_api/test_extract.py::test_extract_mex_person_items` still fails. Can you set that to skip, too?

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -35,9 +35,10 @@ repos:
         additional_dependencies:
           - "backoff>=2.2.1,<3"
           - "click>=8.1.7,<9"
+          - "jinja2>=3.1.2,<4"
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply .pre-commit-config.yaml:38]: pre-commit gives me this error `site-packages\pandas-stubs\io\formats\style_render.pyi:11: error: Cannot find implementation or library stub for module named "jinja2"  [import-not-found]` otherwise

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -35,9 +35,10 @@ repos:
...
           - "pandas-stubs>=2.1.1,<3"
           - "pydantic-settings>=2.1.0,<3"
-          - "pydantic>=2.5.2,<3"
+          - "pydantic[email]>=2.5.2,<3"
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply .pre-commit-config.yaml:41]: pre-commit gives me this error `site-packages\pydantic\networks.py:18: error: Cannot find implementation or library stub for module named "email_validator"  [import-not-found]` otherwise

File: tests/public_api/test_connector.py (hunk 11 above)
Comment (cutoffthetop) [reply tests/public_api/test_connector.py]: added a more general reason here 36ebc12
(also for test_extract_mex_person_items)

File: mex/common/models/resource.py [hunk 15]
```diff
@@ -133,12 +109,18 @@ class BaseResource(BaseModel):
...
-    ] = Field(..., min_length=1)
-    title: list[Text] = Field(..., min_length=1)
-    unitInCharge: list[OrganizationalUnitID] = Field(..., min_length=1)
+    temporal: Timestamp | Annotated[
+        str,
+        Field(
+            examples=["2022-01 bis 2022-03", "Sommer 2023", "nach 2013", "1998-2008"]
+        ),
+    ] | None = None
+    theme: Annotated[
```
[the remark this answers is not in the document]
Comment (cutoffthetop) [reply mex/common/models/resource.py:118]: is this a show-stopper `oh no` or can we fix it? 😅

**Comment by cutoffthetop** [comment]
> Extractor tests fail with this branch. Are you fixing that in this ticket or will you create a separate?

coming up here https://github.com/robert-koch-institut/mex-extractors/pull/111

File: mex/common/models/resource.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rababerladuseladim) [reply mex/common/models/resource.py:118]: not a show stopper but it might reveal how shaky the functions for getting inner types are.

File: .pre-commit-config.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (rababerladuseladim) [reply .pre-commit-config.yaml:41]: alrighty, will be gone with MX-1515 anyway.

File: .pre-commit-config.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (rababerladuseladim) [reply .pre-commit-config.yaml:38]: alrighty, will be gone with MX-1515 anyway.

**Review by rababerladuseladim — APPROVED** [review]
Biu-Ti-Ful!
